### PR TITLE
Enable Eagle3 speculative decoding for GPT-OSS model

### PR DESCRIPTION
**Pull request by eldarkurtic**
This PR adds support for EAGLE3 speculative decoding for GPT-OSS model. Changes tested with a locally trained speculator model, and observed reasonable acceptance rates.

### CODE REVIEW

File: vllm/model_executor/models/gpt_oss.py [hunk 1]
```diff
@@ -239,6 +239,7 @@ def __init__(
         self.make_empty_intermediate_tensors = (
             make_empty_intermediate_tensors_factory(
                 ["hidden_states", "residual"], self.config.hidden_size))
+        self.aux_hidden_state_layers = tuple[int, ...]()
```
Comment (rahul-tuli) [inline vllm/model_executor/models/gpt_oss.py:242]: Should this be:
```suggestion
        self.aux_hidden_state_layers: tuple[int, ...] = ()
```

**Review by rahul-tuli — APPROVED** [review]
Looks good to me!

File: vllm/model_executor/models/gpt_oss.py (hunk 1 above)
Comment (NickLucche) [reply vllm/model_executor/models/gpt_oss.py:242]: it's a newer syntax, it's still consistent

File: vllm/config/speculative.py [hunk 2]
```diff
@@ -527,7 +527,7 @@ def _verify_args(self) -> Self:
                              "speculative decoding is > 1, but got "
                              f"{self.disable_by_batch_size=}")
 
-        eagle3_target_supported = ["llama", "qwen"]
+        eagle3_target_supported = ["llama", "qwen", "gpt_oss"]
```
Comment (NickLucche) [inline vllm/config/speculative.py:530]: unrelated to this PR, but I wonder why do we have to list models here instead of relying on `SupportsEagle3` dispatching

File: vllm/v1/spec_decode/eagle.py [hunk 3]
```diff
@@ -826,15 +826,29 @@ def load_model(self, target_model: nn.Module) -> None:
...
+                    "attribute")
+
+            # Check if shapes match and we found the embedding
+            eagle_shape = self.model.model.embed_tokens.weight.shape
+            target_shape = target_embed_tokens.weight.shape
+            if eagle_shape == target_shape:
+                logger.info(
+                    "Assuming the EAGLE head shares the same vocab embedding"
+                    " with the target model.")
+                del self.model.model.embed_tokens
```
Comment (NickLucche) [inline vllm/v1/spec_decode/eagle.py:846]: should be picked up by gc

**Review by NickLucche — APPROVED** [review]
LGTM but let's wait for someone more familiar with eagle perhaps.

Would be great to have smoke tests for model init at least with layer-capped versions..

**Comment by benchislett** [comment]
Noting #23596, which implements similarly but includes blackwell support via FlashInfer. Should be fine to merge this one first, but we should make sure they're consistent as we will need blackwell support as well

**Comment by jiahanc** [comment]
I think for models with alternative attentions like gpt-oss,  you need to find correct attention builders for draft model (draft model is full attention) like https://github.com/vllm-project/vllm/pull/23596/files#diff-a4809a837fbf535a8f0999b11087a53ec1c53948b50c0a1fe64396bc86de9461R883-R906 . Without this, it will use the sliding window attention, which will have accuracy issues (target model KV will be overwritten by draft model)
Also this part is required https://github.com/vllm-project/vllm/pull/23596/files#diff-80ee7e2a62f9dcfbb8a312dc4e3948557e97ef187290daebbcae1e28596bda29R1107-R1111.

File: vllm/model_executor/models/gpt_oss.py (hunk 1 above)
Comment (mgoin) [reply vllm/model_executor/models/gpt_oss.py:242]: Is it valid for python 3.9? That is still our minimum

Comment (eldarkurtic) [reply vllm/model_executor/models/gpt_oss.py:242]: It is copy-pasted from llama.py and qwen.py in vllm main

File: vllm/config/speculative.py (hunk 2 above)
Comment (eldarkurtic) [reply vllm/config/speculative.py:530]: To be honest not sure, I just followed llama.py / qwen.py to see how eagle is enabled there

File: vllm/v1/spec_decode/eagle.py (hunk 3 above)
Comment (eldarkurtic) [reply vllm/v1/spec_decode/eagle.py:846]: I took this from llama.py and qwen.py. Should I remove it here or leave it for consistency?

**Comment by eldarkurtic** [comment]
The goal of this one was to first add support for the simplest Llama-like-speculator from Eagle3. And then we can build on top of it for more complex archs. This is unblock some preliminary experimentation with speculative decoding.
